File: src/mvc/cpp/tree_model/clickable_tree_view.cpp
#include "clickable_tree_view.h"
#include <QAction>
#include <QMenu>

ClickableTreeView::ClickableTreeView(QWidget *parent) : QTreeView(parent) {

  setContextMenuPolicy(Qt::CustomContextMenu);
  connect(this, &ClickableTreeView::customContextMenuRequested, this,
          [&](const QPoint &pos) {
            QModelIndex index = indexAt(pos);
            if (!index.isValid())
              return;
            QMenu menu;
            auto addChildAction = menu.addAction("Add Child");
            auto removeItemAction = menu.addAction("Remove Item");
            auto editItemAction = menu.addAction("Edit Item");
            auto selectedAction = menu.exec(mapToGlobal(pos));
            switch (menu.actions().indexOf(selectedAction)) {
            case 0:
              addChild(index);
              break;
            case 1:
              removeItem(index);
              break;
            case 2:
              editItem(index);
              break;
            default:
              break;
            }
          });

  setEditTriggers(QAbstractItemView::NoEditTriggers);
}

ClickableTreeView::~ClickableTreeView() {}

TreeModel *ClickableTreeView::model() const {
  return static_cast<TreeModel *>(QTreeView::model());
}

void ClickableTreeView::setModel(TreeModel *model) {
  QTreeView::setModel(model);
}

void ClickableTreeView::addChild(QModelIndex index) {
  auto model = this->model();
  model->itemFromIndex(index)->appendChild(
      new TreeItem(QStringList() << "New Item", model->itemFromIndex(index)));
  expandAll();
}

void ClickableTreeView::removeItem(QModelIndex index) {
  auto model = this->model();
  model->itemFromIndex(index)->parent()->removeChild(
      model->itemFromIndex(index));
  expandAll();
}

void ClickableTreeView::editItem(QModelIndex index) { edit(index); }

File: src/mvc/python/tree_model/clickable_tree_view.py
from PyQt6.QtCore import Qt
from PyQt6.QtWidgets import QAbstractItemView, QMenu, QTreeView
from tree_model import TreeItem


class ClickableTreeView(QTreeView):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setContextMenuPolicy(Qt.ContextMenuPolicy.CustomContextMenu)
        self.customContextMenuRequested.connect(self.contextMenuEvent)
        self.setEditTriggers(QAbstractItemView.EditTrigger.NoEditTriggers)

    def contextMenuEvent(self, pos):
        index = self.indexAt(pos)
        if not index.isValid():
            return
        menu = QMenu()
        menu.addAction("Add Child")
        menu.addAction("Remove Item")
        menu.addAction("Edit Item")
        selectedAction = menu.exec(self.mapToGlobal(pos))
        if not selectedAction:
            return
        value = menu.actions().index(selectedAction)
        if value == 0:
            self.addChild(index)
        elif value == 1:
            self.removeItem(index)
        elif value == 2:
            self.editItem(index)
        else:
            pass

    def model(self):
        return super().model()

    def setModel(self, model):
        super().setModel(model)

    def addChild(self, index):
        model = self.model()
        model.itemFromIndex(index).append_child(
            TreeItem(["New Item"], model.itemFromIndex(index))
        )
        self.expandAll()

    def removeItem(self, index):
        model = self.model()
        model.removeRow(index.row(), index.parent())

    def editItem(self, index):
        self.edit(index)
